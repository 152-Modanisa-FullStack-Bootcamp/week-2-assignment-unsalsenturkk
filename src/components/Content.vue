<template>
  <div
      id="content"
      class="m-flex
               m-flex-column"
  >
    <input
        id="favoriteButton"
        class="favorite-button"
        :class="{'favorite':this.isFavorite(video.id)}"
        @click="favoriteBtnClicked"
        type="submit"
        value=""
    >
    <div
        id="img"
        class="m-flex
        m-flex-column"
        @mouseover="displayHoverImg"
        @mouseleave="displayCoverImg"
    >
      <img
          v-show="hoverImage === false"
          :src="video.coverImage"
          :alt="video.title"
      >
      <img
          v-show="hoverImage === true"
          :src="video.hoverImage"
          :alt="video.title"
          @click="navigateToFavoritePage"
      >
    </div>

    <div
        id="details"
        class="m-flex"
    >
      <img
          id="avatar"
          :src="video.ownerImage"
      >
      <div
          id="meta"
          class="m-flex
                   m-flex-column"

      >
        <div
            id="title"
        >
          {{ video.title }}
        </div>
        <div
            id="ownerName"
        >
          {{ video.ownerName }}
        </div>
        <div
            id="metadata-line"
            class="m-flex"
        >
           <span
               id="viewCount"
           >
             {{ video.viewCount }} B görüntüleme
           </span>
          <span
              id="separator"
          >
              •
            </span>
          <span
              id="publishDateInMonth"
          >
              {{ video.publishDateInMonth }} ay önce
           </span>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import {mapGetters} from "vuex";

export default {

  name: "Content",
  props: {video: Object},
  data() {
    return {
      hoverImage: false,
    }
  },
  methods: {
    displayHoverImg() {
      this.hoverImage = true;
    },
    displayCoverImg() {
      this.hoverImage = false;
    },
    navigateToFavoritePage() {
      this.$router.push({name: "Watch", query: {id: this.video.id}});
    },
    favoriteBtnClicked() {
      this.$store.dispatch("favoriteStatusChanged", this.video)
    }

  },
  watch: {},
  mounted() {

  },
  computed:{
    ...mapGetters({
      isFavorite : "isFavorite"
    }),
  }
}
</script>

<style scoped>

#content {
  width: 360px;
  height: 200px;
  margin-top: 20px;
  margin-right: 30px;
  margin-left: 30px;
  margin-bottom: 50px;
  position: relative;

}

#img {
  max-width: 99%;
  max-height: 80%;
}


#avatar {
  width: 36px;
  height: 36px;
  border-radius: 50px;
  margin-top: 12px;
  margin-right: 12px;
}

#meta {
  margin-top: 12px;
  text-align: left;
}

#title {
  font-weight: bold;
}

#ownerName, #metadata-line {
  color: grey;
}

#separator {
  margin: 0 4px;
}

.favorite-button {
  position: absolute;
  top: 0;
  right: 0;
  background: url('data:image/svg+xml;utf8,<svg fill="red" height="48" viewBox="0 0 24 24" width="48" xmlns="http://www.w3.org/2000/svg"><path d="M0 0h24v24H0z" fill="none"/><path d="M16.5 3c-1.74 0-3.41.81-4.5 2.09C10.91 3.81 9.24 3 7.5 3 4.42 3 2 5.42 2 8.5c0 3.78 3.4 6.86 8.55 11.54L12 21.35l1.45-1.32C18.6 15.36 22 12.28 22 8.5 22 5.42 19.58 3 16.5 3zm-4.4 15.55l-.1.1-.1-.1C7.14 14.24 4 11.39 4 8.5 4 6.5 5.5 5 7.5 5c1.54 0 3.04.99 3.57 2.36h1.87C13.46 5.99 14.96 5 16.5 5c2 0 3.5 1.5 3.5 3.5 0 2.89-3.14 5.74-7.9 10.05z"/></svg>');
  width: 48px;
  height: 48px;
  cursor: pointer;
  border: none;
  margin: 4px;
  transition: background 300ms ease-in-out;
}

.favorite {
background: url('data:image/svg+xml;utf8,<svg fill="red" height="48" viewBox="0 0 24 24" width="48" xmlns="http://www.w3.org/2000/svg"><path d="M0 0h24v24H0z" fill="none"/><path d="M12 21.35l-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"/></svg>');
}

img {
  width: 100%;
  max-height: 100%;
  cursor: pointer;
}

</style>