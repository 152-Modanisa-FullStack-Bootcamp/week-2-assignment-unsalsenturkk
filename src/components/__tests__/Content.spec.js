import {createLocalVue, shallowMount} from "@vue/test-utils";
import Content from "@/components/Content";
import Vuex from "vuex";
import content from "@/components/Content";

describe("Content.vue", () => {

    it('sanity test ', () => {
        return
    });

    describe("Content unit test", () => {
        describe("exists check", () => {
            it('should content component exists', () => {
                const localVue = createLocalVue();
                localVue.use(Vuex)

                const getters = {
                    isFavorite: () => (videoid) => {
                        return false
                    }
                }

                const video = {
                    "id": 1,
                    "videoAddress": "https://www.youtube.com/watch?v=FXpIoQ_rT_c",
                    "coverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-cover.webp",
                    "hoverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-hover.webp",
                    "title": "Vue.js Course for Beginners [2021 Tutorial]",
                    "viewCount": 254,
                    "publishDateInMonth": 4,
                    "ownerImage": "https://yt3.ggpht.com/ytc/AKedOLTtJvQ1Vfew91vemeLaLdhjOwGx3tTBLlreK_QUyA=s68-c-k-c0x00ffffff-no-rj",
                    "ownerName": "freeCodeCamp.org",
                    "description": "Learn Vue 3 by in this full course. Vue.js is an open-source model–view–view model front end JavaScript framework for building user interfaces and single-page applications."
                }

                const store = new Vuex.Store({
                    getters
                })

                const wrapper = shallowMount(Content, {
                    localVue,
                    store,
                    propsData: {
                        video
                    }
                })

                expect(wrapper.exists()).toBeTruthy()
            });
            it('should render favorite button', () => {
                const localVue = createLocalVue();
                localVue.use(Vuex)

                const getters = {
                    isFavorite: () => (videoid) => {
                        return false
                    }
                }

                const video = {
                    "id": 1,
                    "videoAddress": "https://www.youtube.com/watch?v=FXpIoQ_rT_c",
                    "coverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-cover.webp",
                    "hoverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-hover.webp",
                    "title": "Vue.js Course for Beginners [2021 Tutorial]",
                    "viewCount": 254,
                    "publishDateInMonth": 4,
                    "ownerImage": "https://yt3.ggpht.com/ytc/AKedOLTtJvQ1Vfew91vemeLaLdhjOwGx3tTBLlreK_QUyA=s68-c-k-c0x00ffffff-no-rj",
                    "ownerName": "freeCodeCamp.org",
                    "description": "Learn Vue 3 by in this full course. Vue.js is an open-source model–view–view model front end JavaScript framework for building user interfaces and single-page applications."
                }

                const store = new Vuex.Store({
                    getters
                })

                const wrapper = shallowMount(Content, {
                    localVue,
                    store,
                    propsData: {
                        video
                    }
                })
                const favoriteButton = wrapper.find("#favoriteButton")
                expect(favoriteButton.exists()).toBeTruthy()
                expect(favoriteButton.text()).toEqual("")

            })
            it('should render img', () => {
                const localVue = createLocalVue();
                localVue.use(Vuex)

                const getters = {
                    isFavorite: () => (videoid) => {
                        return false
                    }
                }

                const video = {
                    "id": 1,
                    "videoAddress": "https://www.youtube.com/watch?v=FXpIoQ_rT_c",
                    "coverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-cover.webp",
                    "hoverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-hover.webp",
                    "title": "Vue.js Course for Beginners [2021 Tutorial]",
                    "viewCount": 254,
                    "publishDateInMonth": 4,
                    "ownerImage": "https://yt3.ggpht.com/ytc/AKedOLTtJvQ1Vfew91vemeLaLdhjOwGx3tTBLlreK_QUyA=s68-c-k-c0x00ffffff-no-rj",
                    "ownerName": "freeCodeCamp.org",
                    "description": "Learn Vue 3 by in this full course. Vue.js is an open-source model–view–view model front end JavaScript framework for building user interfaces and single-page applications."
                }

                const store = new Vuex.Store({
                    getters
                })

                const wrapper = shallowMount(Content, {
                    localVue,
                    store,
                    propsData: {
                        video
                    }
                })

                const img = wrapper.find("#img")
                expect(img.exists()).toBeTruthy()

            })
            it('should render details', () => {
                const localVue = createLocalVue();
                localVue.use(Vuex)

                const getters = {
                    isFavorite: () => (videoid) => {
                        return false
                    }
                }

                const video = {
                    "id": 1,
                    "videoAddress": "https://www.youtube.com/watch?v=FXpIoQ_rT_c",
                    "coverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-cover.webp",
                    "hoverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-hover.webp",
                    "title": "Vue.js Course for Beginners [2021 Tutorial]",
                    "viewCount": 254,
                    "publishDateInMonth": 4,
                    "ownerImage": "https://yt3.ggpht.com/ytc/AKedOLTtJvQ1Vfew91vemeLaLdhjOwGx3tTBLlreK_QUyA=s68-c-k-c0x00ffffff-no-rj",
                    "ownerName": "freeCodeCamp.org",
                    "description": "Learn Vue 3 by in this full course. Vue.js is an open-source model–view–view model front end JavaScript framework for building user interfaces and single-page applications."
                }

                const store = new Vuex.Store({
                    getters
                })

                const wrapper = shallowMount(Content, {
                    localVue,
                    store,
                    propsData: {
                        video
                    }
                })

                const details = wrapper.find("#details")
                expect(details.exists()).toBeTruthy()
            })
        })
        describe('check functionality', () => {
            it("check displayHoverImg functionality", () => {
                const localThis = {
                    hoverImage: false,
                }
                content.methods.displayHoverImg.call(localThis)
                expect(localThis.hoverImage).toBeTruthy()
            })
            it("check displayCoverImg functionality", () => {
                const localThis = {
                    hoverImage: false,
                }
                content.methods.displayCoverImg.call(localThis)
                expect(localThis.hoverImage).toBeFalsy()
            })
            it("check navigateToFavoritePage functionality", () => {
                const video = {
                    "id": 1,
                    "videoAddress": "https://www.youtube.com/watch?v=FXpIoQ_rT_c",
                    "coverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-cover.webp",
                    "hoverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-hover.webp",
                    "title": "Vue.js Course for Beginners [2021 Tutorial]",
                    "viewCount": 254,
                    "publishDateInMonth": 4,
                    "ownerImage": "https://yt3.ggpht.com/ytc/AKedOLTtJvQ1Vfew91vemeLaLdhjOwGx3tTBLlreK_QUyA=s68-c-k-c0x00ffffff-no-rj",
                    "ownerName": "freeCodeCamp.org",
                    "description": "Learn Vue 3 by in this full course. Vue.js is an open-source model–view–view model front end JavaScript framework for building user interfaces and single-page applications."
                }
                const localThis = {
                    video,
                    $router: {
                        push: jest.fn()
                    },
                }
                content.methods.navigateToFavoritePage.call(localThis)
                expect(localThis.$router.push).toBeCalled()
            })
            it("check favoriteBtnClicked functionality", () => {
                const video = {
                    "id": 1,
                    "videoAddress": "https://www.youtube.com/watch?v=FXpIoQ_rT_c",
                    "coverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-cover.webp",
                    "hoverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-hover.webp",
                    "title": "Vue.js Course for Beginners [2021 Tutorial]",
                    "viewCount": 254,
                    "publishDateInMonth": 4,
                    "ownerImage": "https://yt3.ggpht.com/ytc/AKedOLTtJvQ1Vfew91vemeLaLdhjOwGx3tTBLlreK_QUyA=s68-c-k-c0x00ffffff-no-rj",
                    "ownerName": "freeCodeCamp.org",
                    "description": "Learn Vue 3 by in this full course. Vue.js is an open-source model–view–view model front end JavaScript framework for building user interfaces and single-page applications."
                }
                const dispatch = jest.fn()
                const localThis = {
                    video,
                    $store: {
                        dispatch
                    },
                }
                content.methods.favoriteBtnClicked.call(localThis)
                expect(dispatch).toHaveBeenCalledWith("favoriteStatusChanged",video)
            })
        })
    })
    describe("Content snapshot test", () => {
        it('should renders content correctly', () => {
            const localVue = createLocalVue();
            localVue.use(Vuex)

            const getters = {
                isFavorite: () => (videoid) => {
                    return false
                }
            }

            const video = {
                "id": 1,
                "videoAddress": "https://www.youtube.com/watch?v=FXpIoQ_rT_c",
                "coverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-cover.webp",
                "hoverImage": "https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-hover.webp",
                "title": "Vue.js Course for Beginners [2021 Tutorial]",
                "viewCount": 254,
                "publishDateInMonth": 4,
                "ownerImage": "https://yt3.ggpht.com/ytc/AKedOLTtJvQ1Vfew91vemeLaLdhjOwGx3tTBLlreK_QUyA=s68-c-k-c0x00ffffff-no-rj",
                "ownerName": "freeCodeCamp.org",
                "description": "Learn Vue 3 by in this full course. Vue.js is an open-source model–view–view model front end JavaScript framework for building user interfaces and single-page applications."
            }

            const store = new Vuex.Store({
                getters
            })

            const wrapper = shallowMount(Content, {
                localVue,
                store,
                propsData: {
                    video
                }
            })

            expect(wrapper.element).toMatchSnapshot()
        });
    })
})