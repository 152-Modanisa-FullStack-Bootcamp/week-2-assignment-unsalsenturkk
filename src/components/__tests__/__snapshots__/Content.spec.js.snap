// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Content.vue Content snapshot test should renders content correctly 1`] = `
<div
  class="m-flex
             m-flex-column"
  id="content"
>
  <input
    class="favorite-button"
    id="favoriteButton"
    type="submit"
    value=""
  />
   
  <div
    class="m-flex
      m-flex-column"
    id="img"
  >
    <img
      alt="Vue.js Course for Beginners [2021 Tutorial]"
      src="https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-cover.webp"
    />
     
    <img
      alt="Vue.js Course for Beginners [2021 Tutorial]"
      src="https://raw.githubusercontent.com/modanisa/bootcamp-video-db/main/video-images/1-hover.webp"
      style="display: none;"
    />
  </div>
   
  <div
    class="m-flex"
    id="details"
  >
    <img
      id="avatar"
      src="https://yt3.ggpht.com/ytc/AKedOLTtJvQ1Vfew91vemeLaLdhjOwGx3tTBLlreK_QUyA=s68-c-k-c0x00ffffff-no-rj"
    />
     
    <div
      class="m-flex
                 m-flex-column"
      id="meta"
    >
      <div
        id="title"
      >
        
        Vue.js Course for Beginners [2021 Tutorial]
      
      </div>
       
      <div
        id="ownerName"
      >
        
        freeCodeCamp.org
      
      </div>
       
      <div
        class="m-flex"
        id="metadata-line"
      >
        <span
          id="viewCount"
        >
          
           254 B görüntüleme
         
        </span>
         
        <span
          id="separator"
        >
          
            •
          
        </span>
         
        <span
          id="publishDateInMonth"
        >
          
            4 ay önce
         
        </span>
      </div>
    </div>
  </div>
</div>
`;
